x/sp/keeper/sp.go

```diff
@@ -120,3 +131,38 @@ func (k Keeper) GetAllStorageProviders(ctx sdk.Context) (sps []types.StorageProv
 	}
 	return sps
 }
+
+func (k Keeper) Exit(ctx sdk.Context, sp *types.StorageProvider) error {
+	store := ctx.KVStore(k.storeKey)
+
+	// send back the total deposit
+	coins := sdk.NewCoins(sdk.NewCoin(k.DepositDenomForSP(ctx), sp.TotalDeposit))
```

Comment:
I think it is a better place to make the token transfer logic in `CompleteStorageProviderExit` function...  To me, a keeper should focus on the state in its scope..

Reply:
refacted.